Newsletters: add 160 (2021-08-04) (#617)
bitcoinops/bitcoinops.github.io: _includes/specials/taproot/en/06-multisignatures.md

-->{% endcomment %}
In the 1,000 blocks received prior to this writing, 11% of all
transaction inputs contained a multisig opcode.  Two of the largest and
most immediate benefits of taproot will manifest if many of the users
and services creating those transactions switch from multisig opcodes to
scriptless [multisignatures][topic multisignature].

The first major benefit will be a reduction in transaction size.
Script-based multisigs increase in size as more keys and signatures are
required, but multisignatures are a constant small size.  The smallest
effective multisig policy (1-of-2) requires more space than a
multisignature policy that can involve thousands of signers.  The
decrease in size leads to direct reduction in fees for the
multisignature users and an indirect reduction in fees for all users as
the same amount of demand for confirmed transactions can be fulfilled
using a smaller amount of block space.

{:.center}
![Plot showing the savings for multisignatures compared to multisig](/img/posts/2021-07-multisignature-savings.png)

The second major benefit is improved privacy.  Each use of multisigs is

> distinctively recorded to the block chain where surveilors

Author: It's been used a few times in mailing list discussions, but some searching indicates a word that's in the dictionary (though I've never heard it used and Firefox's autocorrect doesn't like it) is surveillant.  https://duckduckgo.com/?t=ffab&q=surveillant&ia=definition I'll switch to that.

can use them to make informed guesses about the wallet history and
current balance of individual users.  For example, looking at block
692,039, we can distinguish not just the multisig spends from the
single-sig spends but also distinguish between different set sizes and
thresholds for the multisigs.

{:.center}
![Illustration of the lack of witness fungibility in current blocks](/img/posts/2021-07-multisig-unfungible.png)

By comparison, a third party looking only at block chain data can't tell
that a spender used a multisignature.  When a multisignature is used for
a keypath spend, it is indistinguishable from single-sig spends.  If all
single-sig and multisigs in the block above were switched to P2TR
keypath spends, only a few exotic spends would be distinguishable by
their scripts (and even some of those might be able to sometimes use

> keypath spends).

Reviewer: The 'taproot assumption' is that _all_ outputs can be spent using a multisignature keypath spend:

> there exists some monotone function on plain public keys and nothing else that is sufficient to authorize a transaction

(https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-February/015700.html)

Can we replace "even some of those might be able to use keypath spends" to "and even those could use keypath spends in the best case" or similar.

{:.center}
![Illustration of how fungibile witnesses could be ideally](/img/posts/2021-07-multisignature-fungible.png)

### Using multisignatures

We're aware of three schnorr-based multisignature schemes designed
specifically for Bitcoin, all members of the [MuSig][topic musig]
family:

- **MuSig** (also called MuSig1), which should be simple to implement
  but which requires three rounds of communication during the signing
  process.

> - **MuSig2**, also simple to implement and which eliminates one round of
>   communication and allows another round to be combined with key
>   exchange so that it's possible to use a somewhat similar signing
>   process to what we use today with script-based multisig.  This does

Reviewer: Long sentence could be split up:

```suggestion
- **MuSig2** is also simple to implement. It eliminates one round of
  communication and allows another round to be combined with key
  exchange. That means it's possible to use a somewhat similar signing
  process to what we use today with script-based multisig.  This does
```

  require storing extra data and ensuring your signing software or
  hardware can't be tricked into unknowingly repeating part of the
  signing session.

- **MuSig-DN** (Deterministic Nonce), significantly more complex to
  implement and its communication between participants can't be combined

>   with key exchange.  However, it's not vulnerable to the repeated

Reviewer: This sentence could also be split up.

  session attack.

All signers have to agree on the protocol to use, so there will likely
be a network effect where almost all implementations use the same

> protocol.  The authors of the MuSig proposals [suggest][nick ruffing

Reviewer: Maybe, although I think we're a very long way off being able to create generalised multisignatures across an arbitrary set of wallet implementations. I would have thought that most people using multisignature initially would just use a single wallet for all of the signers, which would mean that they could choose any multisignature scheme they want.

Author: > I think we're a very long way off being able to create generalised multisignatures across an arbitrary set of wallet implementations.

Interesting!  I thought the opposite.  All the work projects like HWI have put into creating compatibility shims, plus the need for only small upgrades to PSBT, seems to me to imply that all that's left to do is for actual signing hardware/software to implement MuSigWhatever.  Surely that's two weeks work at most 😀

> I would have thought that most people using multisignature initially would just use a single wallet for all of the signers, which would mean that they could choose any multisignature scheme they want.

That probably is true.  But network effects travel back through time through the anticipation of those effects.  I think there are few wallet developers who want to spend their time implementing a different protocol than the protocol they expect to become widely used in the long term.

Anyway, I'll see if I can tone down the statement in the text.

Reviewer: New wording looks good to me. Thanks!

utility.  <!-- "[...] there is no reason to prefer MuSig1 over MuSig2
[...] we expect that most applications will choose MuSig2 over MuSig-DN
[...]" -->

There's an open and actively-developed [PR][-zkp 131] to the libsecp256k1-zkp
project to add MuSig2 support.  We expect the basic workflow will look
something like the following:

1. The wallet for each participant generates a [BIP32][] xpub that is shared
   with all the other participants (the same as is commonly done now for

>    multisigs)

Reviewer 2: s/multisigs)/multisigs).

2. The wallet also generates a set of nonces that are also shared with
   the other participants.  The wallet can generate these nonces using
   BIP32 hardened derivation.  Nonces are 32 bytes and you need two of
   them per signature.  For infrequently used wallets, all the nonces
   needed for the entire wallet lifetime can be shared up front.  For
   more frequently used wallets (e.g. LN routing nodes), each wallet can
   send its signature for the current transaction along with its nonces
   for the next transaction.

3. Any of the wallets can then generate an aggregated public key by

>    combining its pubkey at a certain BIP32 depth with pubkeys at the

Reviewer: @Xekyo do you suggest any change to the wording here? I think this is probably fine as it is, and talking about descriptors here probably distracts from the more important concept.

   The partial signatures are then combined to create the final
   signature and the transaction is broadcast.

### Threshold signing

By themselves, the MuSig family of multisignature schemes only give you
n-of-n signing---every party who contributes a key towards the
aggregated public key must also contribute a partial signature to the
final signature.  That works perfectly well as a direct replacement for
some uses of script-based multisig today, such as spending 2-of-2 LN
funding outputs, but it's a departure from other popular policies such
as the 2-of-3 multisig script used by many exchanges.

Several developers are working [threshold signature][topic threshold
signature] schemes that will bring the same efficiency and privacy
benefits of multisignatures to k-of-n scenarios, but there's a simple
trick that can be used until those schemes are available.

In many threshold cases, it's known in advance which participants are
most likely to sign.  For example, in a 2-of-3 situation, it might be
known that normally Alice and Bob will co-sign, while Carol only signs
if one of the others is unavailable.  For this set of circumstances, the

> primary keys can use a multisignature for the taproot key-path spend (e.g.

Reviewer: ```suggestion
primary keys can use a multisignature for the taproot keypath spend (e.g.
```

Bob and Carol) can use multisignatures with the `OP_CHECKSIG` opcode in

> separate branches of a [tapscript][topic tapscript].

Reviewer 2: maybe something like s/separate branches of a tapscript./separate branches of a tree of tapscripts.

edit: I think I see what you mean. Using a single tapscript with OP_CHECKSIGs instead of 2 tapscript leaves in the taproot tree. If so maybe the word branches through me off.

Author: Agree with suggestion, will fix.  FYI, I was thinking of two tapscript leaves.  The whole thing is:

       A+B
       /  \
    A+C   B+C

> separate branches of a [tapscript][topic tapscript].

Reviewer: 'leaves' seems slightly better to me, but I think both are fine.

separate branches of a [tapscript][topic tapscript].

In the normal case, the above has exactly as much efficiency and privacy
as a single-sig or multisignature transaction.  In the abnormal case,

> spending still works as expected and remains relatively efficient.

Reviewer: Maybe:

```suggestion
spending still works as expected and remains relatively efficient and private.
```

Using a OP_CHECKSIG in a scriptpath is still more private than using OP_CHECKMULTISIG, since it doesn't reveal how many signers there are.
